### PR TITLE
Workspace.image_from_*: allow filtering by filename

### PR DESCRIPTION
**Pull request by bertsky**
Motivation for this: since we have deprecated `Workspace.resolve_image_as_pil` (because it cannot calculate the coordinate transform that would belong to the image; and we don't want processors to search for AlternativeImages from scratch), we lost a way to deal with the use-case that the caller does _not_ want to create new images dynamically, but instead already knows the desired AlternativeImage (and therefore its path).

Here we therefore (re)introduce that possibility with a kwarg `filename` for the image retrieval functions. (So the difference is that now you have to pass an image path that belongs to an existing AlternativeImage, but then you do get the coordinate transform along with the image.)

@kba I can see no harm here and have already tested locally.

### CODE REVIEW

**Review by kba — APPROVED** [review]
Makes sense and since it's just an additional filter, I see no issue with compatibility.

File: ocrd/ocrd/workspace.py [hunk 1]
```diff
@@ -688,6 +695,10 @@ def image_from_page(self, page, page_id,
                     fill=fill, transparency=transparency)
 
         # verify constraints again:
+        if filename and not getattr(page_image, 'filename', '').endswith(filename):
```
Comment (hnesk) [inline ocrd/ocrd/workspace.py:698]: page_image.filename is always empty for me, so this check always fails

Comment (bertsky) [reply ocrd/ocrd/workspace.py:698]: > page_image.filename is always empty for me, so this check always fails

then that AlternativeImage is not **maximal**, i.e. there are operations (derotation, cropping) that are possible from the description of the segment but have not been applied to the image yet – so the API does it automatically before returning an image. Any PIL.Image operation will invalidate the `filename` attribute, though.

Can you try retrieving the image with additional filters like `feature_filter=deskewed`?

Could you share the snipped of the segment/page that shows the AlternativeImage elements with their comments?

Comment (bertsky) [reply ocrd/ocrd/workspace.py:698]: I agree it's somewhat unexpected, but we have no other choice from the API point of view. Other callers would be surprised to _not_ get an image that corresponds with the features of the segment.

But we could clarify that in the docstring...

**Pull request merged**